File: pat/trigger_PAT.py
#import required libraries
from pathlib import Path
import subprocess
import textwrap
import itertools
import sys
#define global variables
TEMPLATE_PATH = Path("template.csp") 
PAT_EXE = Path(r"MONO-PAT-v3.6.0/PAT3.Console.exe") 
OUTPUT_FILE = Path("output.txt")
RUN_CSP = Path("run.csp")

#define function to build CSP from template CSP 
def build_csp(grid, start_rc, goal_rc, M, out_path="run.csp"):
    #get positions
    start_r, start_c = start_rc
    goal_r,  goal_c  = goal_rc
    #get template text
    template = TEMPLATE_PATH.read_text()
    #replace place holders with updated parameters
    filled = (template
              .replace("{{M}}", str(M))
              .replace("{{GRID}}", str(grid))
              .replace("{{START_R}}", str(start_r))
              .replace("{{START_C}}", str(start_c))
              .replace("{{GOAL_R}}", str(goal_r))
              .replace("{{GOAL_C}}", str(goal_c)))
    #write to output file
    out_file = Path(out_path)
    out_file.write_text(filled)
    return out_file

#run command line call to PAT
def run_pat(csp_path):
    #run subprocess
    print(str(PAT_EXE), str(csp_path), str(OUTPUT_FILE))
    result = subprocess.run(['mono', str(PAT_EXE), str(csp_path), str(OUTPUT_FILE)],text=True, capture_output=True)
    print(f"result: {result}") 
    if result.returncode != 0:
        raise RuntimeError(f"PAT exited with {result.returncode}")

#main code
if __name__ == "__main__":
    # --- example data ----------------------------------------------------
    grid = [
            1, 1, 1,-1, 1, 1, 1, 1,
            1, 1, 1,-1,-1, 1, 1, 1,
            1,-1,-1, 1, 1, 1, 1, 1,
            1,-1, 1, 1, 1, 1, 1,-1,
            1, 1, 1, 1, 1, 1,-1,-1,
            1, 1, 1, 1, 1,-1, 1,-1,
            1, 1, 1, 1, 1,-1, 1, 1,
            1, 1, 1, 1, 1,-1, 1, 1,
    ]
    start_pos = (4, 1)
    goal_pos  = (0, 4)
    grid_size = 8
    # ---------------------------------------------------------------------

    #get CSP file and run PAT 
    csp_file = build_csp(grid, start_pos, goal_pos, grid_size) 
    run_pat(csp_file)

File: pat/trigger_PAT.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <filesystem>
#include <cstdlib>
#include <sstream>

namespace fs = std::filesystem;

// Global constants
const fs::path TEMPLATE_PATH = "../template.csp";
const fs::path PAT_EXE = "../MONO-PAT-v3.6.0/PAT3.Console.exe";
const fs::path OUTPUT_FILE = "../output.txt";
const fs::path RUN_CSP = "../run.csp";

// Utility function to replace all occurrences of a substring
std::string replace_all(std::string str, const std::string& from, const std::string& to) {
    size_t start_pos = 0;
    while((start_pos = str.find(from, start_pos)) != std::string::npos) {
        str.replace(start_pos, from.length(), to);
        start_pos += to.length();
    }
    return str;
}

// Function to build CSP from template CSP
fs::path build_csp(const std::vector<int>& grid, 
                   std::pair<int, int> start_rc,
                   std::pair<int, int> goal_rc,
                   int M,
                   const std::string& out_path) {
    // Get positions
    auto [start_r, start_c] = start_rc;
    auto [goal_r, goal_c] = goal_rc;

    // Read template text
    std::ifstream template_file(TEMPLATE_PATH);
    if (!template_file.is_open()) {
        throw std::runtime_error("Could not open template file");
    }

    std::stringstream buffer;
    buffer << template_file.rdbuf();
    std::string template_content = buffer.str();

    // Convert grid to string
    std::stringstream grid_str;
    grid_str << "[";
    for (size_t i = 0; i < grid.size(); ++i) {
        grid_str << grid[i];
        if (i < grid.size() - 1) grid_str << ",";
    }
    grid_str << "]";

    // Replace placeholders with updated parameters
    std::string filled = template_content;
    filled = replace_all(filled, "{{M}}", std::to_string(M));
    filled = replace_all(filled, "{{GRID}}", grid_str.str());
    filled = replace_all(filled, "{{START_R}}", std::to_string(start_r));
    filled = replace_all(filled, "{{START_C}}", std::to_string(start_c));
    filled = replace_all(filled, "{{GOAL_R}}", std::to_string(goal_r));
    filled = replace_all(filled, "{{GOAL_C}}", std::to_string(goal_c));

    // Write to output file
    fs::path out_file = out_path;
    std::ofstream output(out_file);
    if (!output.is_open()) {
        throw std::runtime_error("Could not open output file for writing");
    }
    output << filled;
    output.close();

    return out_file;
}

// Function to run PAT
void run_pat(const fs::path& csp_path) {
    std::cout << PAT_EXE.string() << " " << csp_path.string() << " " << OUTPUT_FILE.string() << std::endl;
    
    // Construct command
    std::string command = "mono " + PAT_EXE.string() + " " + csp_path.string() + " " + OUTPUT_FILE.string();
    
    // Execute command
    int return_code = system(command.c_str());
    
    if (return_code != 0) {
        throw std::runtime_error("PAT exited with error code: " + std::to_string(return_code));
    }

    // Read and print verification result
    if (fs::exists(OUTPUT_FILE)) {
        std::ifstream result_file(OUTPUT_FILE);
        if (result_file.is_open()) {
            std::string line;
            std::cout << "Verification Result:" << std::endl;
            while (std::getline(result_file, line)) {
                std::cout << line << std::endl;
            }
        }
    }
}

int main() {
    // Example data
    std::vector<int> grid = {
        1, 1, 1,-1, 1, 1, 1, 1,
        1, 1, 1,-1,-1, 1, 1, 1,
        1,-1,-1, 1, 1, 1, 1, 1,
        1,-1, 1, 1, 1, 1, 1,-1,
        1, 1, 1, 1, 1, 1,-1,-1,
        1, 1, 1, 1, 1,-1, 1,-1,
        1, 1, 1, 1, 1,-1, 1, 1,
        1, 1, 1, 1, 1,-1, 1, 1
    };

    std::pair<int, int> start_pos = {4, 1};
    std::pair<int, int> goal_pos = {0, 4};
    int grid_size = 8;

    try {
        // Get CSP file and run PAT
        fs::path csp_file = build_csp(grid, start_pos, goal_pos, grid_size, RUN_CSP.string());
        run_pat(csp_file);
    }
    catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
